### PR TITLE
JPC: Show Jetpack plans first

### PR DESCRIPTION
**Pull request by johnHackworth**
Something we want to try is to check how the results change if we show the jetpack plans selection screen to the users before the authorization, during the jetpack connect process, instead of doing it once the site is fully connected.

This PR implements this test.

image

untitled screencast 18

How to test
========

1. Open http://calypso.localhost:3000/jetpack/connect
2. In your development console, write `localStorage.setItem('ABTests','{"jetpackConnectPlansFirst_20160715":"showPlansBeforeAuth"}')`
3. Connect one of your (disconnected) jetpack sites. After you enter your url, the next step you should see is the plans selection screen
4. Once you select a plan, the connection process should start
5. Once your site is connected, you should be redirected to the checkout with the plan you selected already in the cart
5b. If you selected the free plan, you should be redirected to /posts instead

ping @rickybanister @oskosk @roccotripaldi @ryelle & @richardmuscat 

Test live: https://calypso.live/?branch=add/jetpack-plans-first

### CODE REVIEW

**Comment by roccotripaldi** [comment]
rebased to fix conflicts from #6868

**Comment by roccotripaldi** [comment]
rebased again to fix conflicts on active-tests.js

**Comment by roccotripaldi** [comment]
I'm going to try these flows, and post my findings:
- Site that does not have Jetpack installed
- Site that has an inactive Jetpack
- Site that has an unconnected Jetpack
- Site that has a connected Jetpack with no plan
- Site that has an unconnected Jetpack with no plan
- Site that has an unconnected Jetpack that already has a plan

**Comment by roccotripaldi** [comment]
Woohoo! It was pretty dang smooth on a site with no Jetpack. 
Here is a quick screen cast. Note, I was logged into wpcom, and i had to inject a little magic to ensure I finished up in calypso.localhost

https://cldup.com/slcutXn5oZ.mov

**Comment by roccotripaldi** [comment]
Also a smooth experience on a site with an inactive Jetpack. This time I selected the Free plan, and landed in Posts as expected.

**Comment by roccotripaldi** [comment]
Also smooth on an unconnected Jetpack site.

**Comment by roccotripaldi** [comment]
Here's is what happened on an already connected site that has no plan. Not ideal, and we'll have to come up with a better way to handle this.
https://cloudup.com/chKoU9ioJhV

**Comment by roccotripaldi** [comment]
Here is a screen cast of an unconnected site that already has a plan:
https://cloudup.com/cJqlKFSYUPk
It ends well, but it does present the plan selection screen, and that screen is completely unaware that the site already has a premium plan.

**Comment by roccotripaldi** [comment]
I think there is enough to chew on for now. I also want to test logged out flows, but I don't want to overwhelm you with too much feedback. We'll do logged out later.

**Comment by johnHackworth** [comment]
@roccotripaldi  now it shows the 'connected by another user' directly, without having to wait to select a plan first:

https://cloudup.com/c00lDJGvZtW

Regardings the other bug ...  the problem here is that when we select the plan first, the site is still not connected, so we don't have any info about if it already have a plan or not. I think this case would need a new endpoint to fetch that info, so I would add it in another PR, if you agree

**Comment by roccotripaldi** [comment]
I rebased and tested all logged out flows. Everything is pretty smooth!
The only thing is, I always get the plans-first variation, even if I don't set localStorage.
I see that it is a 50/50 chance. Maybe I am just lucky :)

File: client/signup/jetpack-connect/authorize-form.jsx [hunk 1]
```diff
@@ -492,19 +501,43 @@ const JetpackConnectAuthorizeForm = React.createClass( {
...
+	},
+
+	render() {
+		const { queryObject } = this.props.jetpackConnectAuthorize;
+
+		if ( queryObject.already_authorized && ! this.props.isAlreadyOnSitesList ) {
+			this.renderMainForm();
+		}
+
+		if ( this.props.plansFirst && ! this.props.hasJetpackPlanSelected ) {
```
Comment (roccotripaldi) [inline client/signup/jetpack-connect/authorize-form.jsx]: I think this needs to be `'showPlansBeforeAuth' === this.props.plansFirst`

**Comment by roccotripaldi** [comment]
> Maybe I am just lucky :)

Nop, this PR will always be plans-first. I left a code comment.

**Comment by rickybanister** [comment]
@johnHackworth do you know how close the new plans page is to being ready for jetpack plans? If it's ready, I'd suggest we skip this as a test and move over to the new plans page as the first step in the process. I'm afraid we'll just have to rerun this once the new page is ready..

**Comment by roccotripaldi** [comment]
@rickybanister the new plans is pretty close to being live in the Jetpack connect flows. we could merge it today.  once #7165 is in, is there any reason we cannot merge this to AB test Plans First v. Plans Last?  In both cases, the new plans layout will be used.

**Comment by rickybanister** [comment]
Cool, no reason why we couldn't merge them. Testing apples/apples would be my preference.

**Comment by johnHackworth** [comment]
Good catch, @roccotripaldi ! I fixed it in a new commit

**Comment by roccotripaldi** [comment]
Nice. Looks good!

**Comment by johnHackworth** [comment]
This has been painful, but I finally have the new plans look merged here and everything seems to be working. But I really need someone else to test this too, since I don't trust at all that I'm covering all the cases and that this huge merge has not broke something.

untitled screencast 16

**Comment by johnHackworth** [comment]
I've updated this with the latest master.

@roccotripaldi  @rickybanister  can we just merge this in? If we don't want to launch the test already, we can set it up so the a/b test always choose the plans-last version ... but as long as we keep this PR unmerged, I keep needing to come back to it to spend a couple of hours resolving conflicts with current `master` so it's up to date.

**Comment by roccotripaldi** [comment]
After an other round of testing, this seems good to me. I think we can merge it if we put the appropriate logic in to ensure folks will not see plans first yet. And we can switch in on in an other PR.

**Comment by rickybanister** [comment]
Makes sense to me

**Comment by johnHackworth** [comment]
closing this and continuing the work in https://github.com/Automattic/wp-calypso/pull/8920

**Pull request closed without merging**